packages/blob-storage/internal/indexer/indexer.go

```diff
@@ -0,0 +1,94 @@
+package indexer
+
+import (
+	"context"
+	"log"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// Indexer struct holds the configuration and state for the Ethereum chain listener.
+type Indexer struct {
+	rpcURL            string
+	beaconURL         string
+	networkName       string
+	contractAddress   common.Address
+	eventHash         common.Hash
+	startHeight       *big.Int
+	eventCallbackFunc func(string, string, string, types.Log)
+}
+
+// NewIndexer creates a new Indexer instance.
+func NewIndexer(rpcURL, beaconURL, networkName, contractAddress string, pastEvents bool, startBlockNumber *big.Int) *Indexer {
+	client, err := ethclient.Dial(rpcURL)
+	if err != nil {
+		log.Fatal("Failed to connect to the Ethereum client:", err)
+	}
+
+	if pastEvents == false {
+		header, err := client.HeaderByNumber(context.Background(), nil)
+		if err != nil {
+			log.Fatal(err)
+		}
+		startBlockNumber = header.Number
+	}
+
+	slog.Info(client.Client())
+
+	return &Indexer{
+		rpcURL:          rpcURL,
+		beaconURL:       beaconURL,
+		networkName:     networkName,
+		contractAddress: common.HexToAddress(contractAddress),
+		startHeight:     startBlockNumber,
+	}
+}
+
+// SubscribeEvent subscribes to the specified event.
+func (indexer *Indexer) SubscribeEvent(eventHash string, callbackFunc func(string, string, string, types.Log)) {
+	indexer.eventHash = common.HexToHash(eventHash)
+	indexer.eventCallbackFunc = callbackFunc
+}
+
+// Start starts the Ethereum chain listener/indexer.
+func (indexer *Indexer) Start() {
+	client, err := ethclient.Dial(indexer.rpcURL)
+	if err != nil {
+		log.Fatal("Failed to connect to the Ethereum client:", err)
+	}
+
+	query := ethereum.FilterQuery{
+		Addresses: []common.Address{indexer.contractAddress},
+		FromBlock: indexer.startHeight,
+	}
+
+	logsCh := make(chan types.Log)
+	sub, err := client.SubscribeFilterLogs(context.Background(), query, logsCh)
```

Comment:
Geth subscription is not promising: https://geth.ethereum.org/docs/interacting-with-geth/rpc/pubsub#create-subscriptions 

<img width="782" alt="image" src="https://github.com/taikoxyz/taiko-mono/assets/104078303/051b19f1-15a5-4306-9af5-836f6b355d32">

Maybe we need to scan all L1 blocks one by one, and we have the same need in taiko-client, so not 100% sure but probly maybe u can reuse this package: https://github.com/taikoxyz/taiko-client/blob/main/pkg/chain_iterator/event_iterator/block_proposed_iterator.go

Reply:
TBH i'd leave it to @cyberhorsey to not waste to much time with theses. I think it would take a pro-go dev. `x` time, but for me maybe `2x` or even `3x`

Reply:
done, we filter on an event loop in batches now